### PR TITLE
refactor: gateway stats

### PR DESCRIPTION
**Pull request by Sidddddarth**
# Description

Using a custom composite struct `SourceStat` instead of a nested map in `userWebRequestWorkerProcess` for capturing statistics.

**Refactoring Notes:**

- `gateway.go#userWebRequestWorkerProcess`
   - Decoupled job extraction logic to a separate method, `getJobFromRequest`.
   - Collecting request statistics is performed exclusively inside `userWebRequestWorkerProcess` and statistics for the whole batch are reported once, at the end of the method.
   - Getting rid of a handful of intermediate maps used for capturing statistics. Statistics collection is now encapsulated in `SourceStat` struct.
- `webhook.go#RequestHandler`
   - Request statistics are collected and reported exclusively in `RequestHandler` method. Previously this was done only in `failRequest`.
   - Collecting successful webhook request statistics.

## Notion Ticket

https://www.notion.so/rudderstacks/Refactor-gateway-stats-82ff63ac487341f3ac8b795df39c2f04

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: gateway/gateway.go [hunk 1]
```diff
@@ -446,19 +433,12 @@ func (gateway *HandleT) getSourceTagFromWriteKey(writeKey string) string {
 // Finally sends responses(error) if any back to the webRequests over their `done` channels
 func (gateway *HandleT) userWebRequestWorkerProcess(userWebRequestWorker *userWebRequestWorkerT) {
 	for breq := range userWebRequestWorker.batchRequestQ {
-		counter := atomic.AddUint64(&gateway.webRequestBatchCount, 1)
+		// counter := atomic.AddUint64(&gateway.webRequestBatchCount, 1)
```
Comment (atzoum) [inline gateway/gateway.go:436]: why is it commented out?

File: gateway/webhook/webhook.go [hunk 2]
```diff
@@ -253,6 +264,8 @@ func (webhook *HandleT) RequestHandler(w http.ResponseWriter, r *http.Request) {
 	}
 	pkgLogger.Debugf("IP: %s -- %s -- Response: 200, %s", misc.GetIPFromReq(r), r.URL.Path, response.GetStatus(response.Ok))
 	_, _ = w.Write(payload)
+	ss.RequestSucceeded()
```
Comment (atzoum) [inline gateway/webhook/webhook.go:267]: previously we were not reporting successful webhook requests

File: gateway/internal/stats/stats.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
+package stats
+
+import "github.com/rudderlabs/rudder-server/services/stats"
+
+type SourceStat struct {
```
Comment (atzoum) [inline gateway/internal/stats/stats.go:5]: we can fully cover with tests this struct's methods

File: gateway/gateway.go [hunk 4]
```diff
@@ -446,226 +432,43 @@ func (gateway *HandleT) getSourceTagFromWriteKey(writeKey string) string {
...
-			body, _ = sjson.SetBytes(body, "writeKey", writeKey)
-			body, _ = sjson.SetBytes(body, "receivedAt", time.Now().Format(misc.RFC3339Milli))
-			eventBatchesToRecord = append(eventBatchesToRecord, string(body))
-			sourcesJobRunID := gjson.GetBytes(body, "batch.0.context.sources.job_run_id").Str   // pick the job_run_id from the first event of batch. We are assuming job_run_id will be same for all events in a batch and the batch is coming from rudder-sources
-			sourcesTaskRunID := gjson.GetBytes(body, "batch.0.context.sources.task_run_id").Str // pick the task_run_id from the first event of batch. We are assuming task_run_id will be same for all events in a batch and the batch is coming from rudder-sources
-			id := uuid.New()
-
-			params := map[string]interface{}{
-				"source_id":          sourceID,
-				"batch_id":           counter,
```
Comment (atzoum) [inline gateway/gateway.go:645]: @chandumlg do we still need this `batch_id` counter? If yes, for what purpose?

File: gateway/gateway.go [hunk 5]
```diff
@@ -172,7 +173,6 @@ type HandleT struct {
 	trackSuccessCount int
 	trackFailureCount int
 
-	webRequestBatchCount                                       uint64
```
Comment (atzoum) [inline gateway/gateway.go:175]: @chandumlg we removed this global counter which was used for populating the `batch_id` parameter. Something to do with pending events?

**Review by atzoum — APPROVED** [review]

File: gateway/internal/stats/stats.go [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+		"writeKey":    ss.WriteKey,
+		"reqType":     ss.ReqType,
+		"workspaceId": ss.WorkspaceID,
+		"sourceID":    ss.SourceID,
+	}
+	if ss.reason != "" {
+		failedTags["reason"] = ss.reason
+	}
+
+	s.NewTaggedStat("gateway.write_key_requests", stats.CountType, tags).Count(ss.requests.total)
```
[the remark this answers is not in the document]
Comment (atzoum) [reply gateway/internal/stats/stats.go:96]: Sum where? We have dashboard and alerts already that are dependent upon both these measurements (`gateway_write_key_requests` & `gateway_write_key_events`).

[the remark this answers is not in the document]
Comment (atzoum) [reply gateway/internal/stats/stats.go:96]: We should treat this as a separate task. It requires things to happen before that, plus influx dashboards & alerts need to be thrown away (adding up different measurements in influx is not possible)

[the remark this answers is not in the document]
Comment (fracasula) [reply gateway/internal/stats/stats.go:96]: Sounds good to me :+1:

**Review by cisse21 — APPROVED** [review]

**Pull request merged**